### PR TITLE
Fix authorization url building where a URL parameter could be a nil value

### PR DESCRIPTION
**Pull request by chris-at-work**
* original bug seen from production logs
* using URI stdlib over string manipulation
* added tests

### CODE REVIEW

File: spec/smart_app_launch/app_redirect_test_spec.rb [hunk 1]
```diff
@@ -116,4 +116,36 @@ def run(runnable, inputs = {})
...
+    context 'when there are existing URL paramters' do
+      let(:base_url) { 'https://example.com?keep=me' }
+      let(:result) do
+        subject.authorization_url_builder(base_url, new_params)
+      end
+
+      it { expect(result).to eq "https://example.com?keep=me&one=1&two=2" }
+    end
+
+    context 'when a URL paramter value is nil' do
```
Comment (arscan) [inline spec/smart_app_launch/app_redirect_test_spec.rb:141]: Even more minor than a nitpick: `s/paramter/parameter/g` :)

**Review by arscan — APPROVED** [review]
This is a much cleaner approach.  One thing I was trying to do last week but ran out of time was figure out how clients were able to trigger this exception in the first place.  It seems like we may want some kind of assertion ahead of this with a nice error message if a nil value for a key is not allowed.  This will make it so this code doesn't break in an unhelpful way, but I wonder if it just delays the issue or causes our client to then do something very odd (like send a follow-on request with a parameter with no value, which probably isn't valid, and then make it look like our client is wrong when really it was getting bad data to start from the system under test).  FYI @Jammjammjamm if we merge this we might want to add another ticket to track that down.

**Comment by chris-at-work** [comment]
> This is a much cleaner approach. One thing I was trying to do last week but ran out of time was figure out how clients were able to trigger this exception in the first place. It seems like we may want some kind of assertion ahead of this with a nice error message if a nil value for a key is not allowed.

Yeah this is just defense but not validation.  It's covering and testing the nil value case but not nil key (which should be impossible?).  Validation would be something else and I agree.
